ledger-service/http-json/src/main/scala/com/digitalasset/http/Endpoints.scala

```diff
@@ -51,45 +56,80 @@ class Endpoints(
-      httpResponse(createAndExercise(req))
-    case req @ HttpRequest(POST, Uri.Path("/v1/fetch"), _, _, _) => httpResponse(fetch(req))
-    case req @ HttpRequest(GET, Uri.Path("/v1/query"), _, _, _) => httpResponse(retrieveAll(req))
-    case req @ HttpRequest(POST, Uri.Path("/v1/query"), _, _, _) => httpResponse(query(req))
-    case req @ HttpRequest(GET, Uri.Path("/v1/parties"), _, _, _) => httpResponse(allParties(req))
-    case req @ HttpRequest(POST, Uri.Path("/v1/parties"), _, _, _) => httpResponse(parties(req))
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(createAndExercise(req)))
+    case req @ HttpRequest(POST, Uri.Path("/v1/fetch"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(fetch(req)))
+    case req @ HttpRequest(GET, Uri.Path("/v1/query"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(retrieveAll(req)))
+    case req @ HttpRequest(POST, Uri.Path("/v1/query"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(query(req)))
+    case req @ HttpRequest(GET, Uri.Path("/v1/parties"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(allParties(req)))
+    case req @ HttpRequest(POST, Uri.Path("/v1/parties"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(parties(req)))
     case req @ HttpRequest(POST, Uri.Path("/v1/parties/allocate"), _, _, _) =>
-      httpResponse(allocateParty(req))
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(allocateParty(req)))
     case req @ HttpRequest(GET, Uri.Path("/v1/packages"), _, _, _) =>
-      httpResponse(listPackages(req))
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(listPackages(req)))
     // format: off
     case req @ HttpRequest(
           GET, Uri(_, _, Slash(Segment("v1", Slash(Segment("packages", Slash(Segment(packageId, Empty)))))), _, _),
           _, _, _) =>
-      downloadPackage(req, packageId)
+            mkRequestLoggingContext(req)(implicit lc =>
+      downloadPackage(req, packageId))
     // format: on
     case req @ HttpRequest(POST, Uri.Path("/v1/packages"), _, _, _) =>
-      httpResponse(uploadDarFile(req))
-    case HttpRequest(GET, Uri.Path("/livez"), _, _, _) =>
-      Future.successful(HttpResponse(status = StatusCodes.OK))
-    case HttpRequest(GET, Uri.Path("/readyz"), _, _, _) =>
-      healthService.ready().map(_.toHttpResponse)
+      mkRequestLoggingContext(req)(implicit lc => httpResponse(uploadDarFile(req)))
+    case req @ HttpRequest(GET, Uri.Path("/livez"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => {
+        logger.trace("Incoming Request")
+        Future.successful(HttpResponse(status = StatusCodes.OK))
+      })
+    case req @ HttpRequest(GET, Uri.Path("/readyz"), _, _, _) =>
+      mkRequestLoggingContext(req)(implicit lc => {
+        logger.trace("Incoming Request")
+        healthService.ready().map(_.toHttpResponse)
+      })
```

Comment:
I know, many duplicates here. However I decided to do this so in the parent code calling this code no concrete request needs to be handed down as otherwise this would result in composing the functions for each incoming request (at least if I understood the code in `HttpService` correctly).

Reply:
You can eliminate the duplication without changing the type of `all` or the semantics you have defined.

```scala
import scalaz.std.partialFunction._, scalaz.syntax.arrow._

    lazy val all: PartialFunction[HttpRequest, Future[HttpResponse]] = {
      val dispatch: PartialFunction[HttpRequest, LoggingContextOf[...] => Future[HttpResponse]] = {
          case req @ HttpRequest(POST, Uri.Path("/v1/create"), _, _, _) =>
            implicit lc => httpResponse(create(req))
         // other cases
      }
      (dispatch &&& { case req => req }) andThen {
        case (lcFhr, req) => mkRequestLoggingContext(req)(lcFhr)
      }
    }
```